File: programmers/python/level3/합승_택시_요금.py
max = 20_000_000
def solution(n, s, a, b, fares):
    arr = [[0 if i == j else max for j in range(n+1)] for i in range(n+1)]
    for fare in fares:
        f = fare[0]
        t = fare[1]
        c = fare[2]
        arr[f][t] = c
        arr[t][f] = c
    
    for k in range(1,n+1):
        for i in range(1,n+1):
            for j in range(1,n+1):
                if arr[i][k] + arr[k][j] < arr[i][j]:
                    arr[i][j] = arr[i][k] + arr[k][j]
    
    answer = max
    for num in range(1,n+1): # 합승 지점
        cost = arr[s][num] + arr[num][a] + arr[num][b]
        if cost < answer:
            answer = cost
    return answer

print(solution(	7, 3, 4, 1, [[5, 7, 9], [4, 6, 4], [3, 6, 1], [3, 2, 3], [2, 1, 6]]))

File: programmers/js/new2/level3/합승_택시_요금.js
function solution(n, s, a, b, fares) {
    var answer = Number.MAX_SAFE_INTEGER;
    const map = Array.from(Array(n + 1), (_, i) =>
        Array.from(Array(n + 1), (_, j) => (i === j ? 0 : Number.MAX_SAFE_INTEGER))
    );
    fares.forEach((f) => {
        const [from, to, cost] = f;
        map[from][to] = cost;
        map[to][from] = cost;
    });

    for (let k = 1; k <= n; ++k) {
        for (let i = 1; i <= n; ++i) {
            for (let j = 1; j <= n; ++j) {
                map[i][j] = Math.min(map[i][j], map[i][k] + map[k][j]);
            }
        }
    }

    for (let i = 1; i <= n; ++i) {
        answer = Math.min(answer, map[s][i] + map[i][a] + map[i][b]);
    }
    return answer;
}

console.log(
    solution(6, 4, 6, 2, [
        [4, 1, 10],
        [3, 5, 24],
        [5, 6, 2],
        [3, 1, 41],
        [5, 1, 24],
        [4, 6, 50],
        [2, 4, 66],
        [2, 3, 22],
        [1, 6, 25],
    ])
);
